### PR TITLE
[LEOP-287]Apply all customize features to CRA5-Part1

### PR DESCRIPTION
**Pull request by salleyliu**
This PR will merge all add files in react-scripts:

- Add all files in backpack-addons
- Add js.html and css.html in config
- Add getCSSModuleLocalIdentWithProjectName.js in utils

### CODE REVIEW

File: packages/react-scripts/backpack-addons/splitChunks.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  // https://twitter.com/wSokra/status/969633336732905474
+  // https://medium.com/webpack/webpack-4-code-splitting-chunk-graph-and-the-splitchunks-optimization-be739a861366
+  return {
+    splitChunks: bpkReactScriptsConfig.enableAutomaticChunking
+    ? {
+      chunks: 'all',
+      name: isEnvDevelopment,
+      cacheGroups: bpkReactScriptsConfig.vendorsChunkRegex
+        ? {
+            defaultVendors: {
```
Comment (salleyliu) [inline packages/react-scripts/backpack-addons/splitChunks.js:18]: in webpack5: optimization.splitChunks.cacheGroups.vendors → optimization.splitChunks.cacheGroups.defaultVendors

File: packages/react-scripts/backpack-addons/cssModules.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  const cssRegex = /\.css$/;
+  const cssModuleRegex = /\.module\.css$/;
+  const sassRegex = /\.(scss|sass)$/;
+  const sassModuleRegex = /\.module\.(scss|sass)$/;
+
+  switch (regexType) {
+    case 'css':
+      return {
+        and: [cssRegex, () => !cssModulesEnabled],
+        not: [backpackModulesRegex, scopedBackpackModulesRegex],
```
Comment (salleyliu) [inline packages/react-scripts/backpack-addons/cssModules.js:24]: Accroding webpack5 [Rule.exclude](https://webpack.js.org/configuration/module/#ruleexclude) and Rule.include:  If you supply a Rule.exclude option, you cannot also supply a Rule.resource(and, not, or )
So I'm going to change `exclude` here to `not`

File: packages/react-scripts/backpack-addons/cssModules.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+          and: [cssRegex, () => cssModulesEnabled],
+        },
+        {
+          and: [cssRegex, backpackModulesRegex, scopedBackpackModulesRegex],
+        },
+      ];
+    case 'sass':
+      return {
+        and: [sassRegex, () => !cssModulesEnabled],
+        not: [backpackModulesRegex, scopedBackpackModulesRegex],
```
Comment (salleyliu) [inline packages/react-scripts/backpack-addons/cssModules.js:39]: Same above. I'm going to change `exclude` here to `not`.

File: packages/react-scripts/utils/getCSSModuleLocalIdentWithProjectName.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  const localIdent = getCSSModuleLocalIdent(...args);
+
+  // Create an ident based on the file location, class name and Project Name.
+  // Will be unique across projects
+  // Format [ file/folder ]_[ localName ]__[ 5 character hash of CRA ident and project name ]
+  return (
+    localIdent.substring(0, localIdent.length - 5) +
+    createHash(localIdent + projectName)
+  );
+};
```
Comment (jaysonwu991) [inline packages/react-scripts/utils/getCSSModuleLocalIdentWithProjectName.js:22]: Need an empty new line in the end of this file.

File: packages/react-scripts/config/js.html [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<% _.each(htmlWebpackPlugin.files.js, file => { %>
+    <script src="<%= file %>"></script>
+    <% }); %>
```
Comment (jaysonwu991) [inline packages/react-scripts/config/js.html:3]: Need an empty new line in the end of this file.

File: packages/react-scripts/config/css.html [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<% _.each(htmlWebpackPlugin.files.css, file => { %><link
+    rel="stylesheet"
+    href="<%= file %>"
+  />
+  <% }); %>
```
Comment (jaysonwu991) [inline packages/react-scripts/config/css.html:5]: Need an empty new line in the end of this file.

File: packages/react-scripts/backpack-addons/ssr/statusFile.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+  compiler.hooks.invalid.tap('invalid', () => {
+    inProgress(buildPath);
+  });
+  compiler.hooks.done.tap('done', () => {
+    done(buildPath);
+  });
+};
+
+exports.done = done;
```
Comment (jaysonwu991) [inline packages/react-scripts/backpack-addons/ssr/statusFile.js:31]: Need an empty new line in the end of this file.

File: packages/react-scripts/backpack-addons/ssr/forkSsr.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+   err => {
+     if (err) {
+       ui.clear();
+       ui.log(chalk.red('Failed to compile:') + '\n' + (err.message || err));
+       process.exit(1);
+     }
+ 
+     statusFile.done(paths.appBuildSsr);
+   }
+ );
```
Comment (jaysonwu991) [inline packages/react-scripts/backpack-addons/ssr/forkSsr.js:67]: Need an empty new line in the end of this file.

File: packages/react-scripts/backpack-addons/ssr/customWebpackUtils.js [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+      );
+    }
+  });
+
+  return compiler;
+}
+
+module.exports = {
+  createCustomCompiler,
+};
```
Comment (jaysonwu991) [inline packages/react-scripts/backpack-addons/ssr/customWebpackUtils.js:201]: Need an empty new line in the end of this file.

File: packages/react-scripts/backpack-addons/cssModules.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+const getCSSModuleLocalIdent = () => {
+  return require('../utils/getCSSModuleLocalIdentWithProjectName')(
+    appPackageJson.name
+  );
+};
+
+module.exports = {
+  getStyleTestRegexes,
+  getCSSModuleLocalIdent,
+};
```
Comment (jaysonwu991) [inline packages/react-scripts/backpack-addons/cssModules.js:65]: Need an empty new line in the end of this file.

File: packages/react-scripts/backpack-addons/README.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+- Require features in the file(for example, webpack.config.js) where they are needed.
+  - For each line of code that is changed, make sure to add a comment `// #backpack-addon {{featureName}}`
+  - Try to keep the lines of code changed outside of the backpack-addons folder to a minimum
+`require(...../backpack-addons/...)` files instead of writing it inline as much as possible.
+- Add a description of the new feature in the table above the current document.
+
+## How to upgrade from the upstream `facebook/create-react-app` repo
+
+- Replace all of `packages/react-scripts` with the upstream version.
+- Restore `packages/react-scripts/backpack-addons` all features.
```
Comment (jaysonwu991) [inline packages/react-scripts/backpack-addons/README.md:31]:
```suggestion
- Restore all features in `packages/react-scripts/backpack-addons` folder from the old version of BRS.
```

**Review by jaysonwu991 — COMMENTED** [review]
Need to check the formation of each file and make sure it is easy to understand.

File: packages/react-scripts/config/css.html [hunk 12]
```diff
@@ -0,0 +1,6 @@
...
+
```
Comment (jaysonwu991) [inline packages/react-scripts/config/css.html:6]:
```suggestion
```

Remove extra empty line.

File: packages/react-scripts/config/css.html (hunk 6 above)
Comment (jaysonwu991) [inline packages/react-scripts/config/css.html:5]:
```suggestion
  <% }); %>
  
```

Restore empty line in the end of file.

**Review by paultan2021 — APPROVED** [review]
LGTM

**Pull request merged**